In [18]:
import os, random
import keras
from keras.datasets import fashion_mnist
from keras.models import Sequential
from keras.layers import Dense, BatchNormalization, Dropout
from keras.utils import np_utils
import numpy as np
import tensorflow as tf
os.path.expanduser = lambda path: './'

In [19]:
batch_size = 128
num_classes = 10
epochs = 60

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.np_utils.to_categorical(y_train, num_classes)
y_test = keras.utils.np_utils.to_categorical(y_test, num_classes)

60000 train samples
10000 test samples


# Define Model (Drop rate : 0.2, No BN)

In [20]:
# for reproducibility
import random, os
os.environ['PYTHONHASHSEED']='0'
random.seed(123)
np.random.seed(123)
tf.random.set_seed(123)
sess = tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1,    inter_op_parallelism_threads=1,
                                                               allow_soft_placement=True, device_count = {'CPU': 1}))
from tensorflow.python.keras import backend as K
K.set_session(sess)



kernel_initializer='glorot_uniform'
activation_function = 'relu'

with tf.device('/cpu:0'):
    model = Sequential()
    model.add(Dense(512, activation='relu', input_shape=(784,)))
    model.add(Dropout(0.2))
    #model.add(BatchNormalization())
    model.add(Dense(512, activation='relu'))
    model.add(Dropout(0.2))
    #model.add(BatchNormalization())
    model.add(Dense(num_classes, activation='softmax'))

    model.summary()

    model.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_9 (Dense)             (None, 512)               401920    
                                                                 
 dropout_6 (Dropout)         (None, 512)               0         
                                                                 
 dense_10 (Dense)            (None, 512)               262656    
                                                                 
 dropout_7 (Dropout)         (None, 512)               0         
                                                                 
 dense_11 (Dense)            (None, 10)                5130      
                                                                 
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________


# Start Training

In [21]:
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 7s 16ms/step - loss: 1.1964 - accuracy: 0.6194 - val_loss: 0.7336 - val_accuracy: 0.7558
Epoch 2/60
375/375 [==============================] - 6s 16ms/step - loss: 0.7406 - accuracy: 0.7520 - val_loss: 0.6155 - val_accuracy: 0.7909
Epoch 3/60
375/375 [==============================] - 5s 13ms/step - loss: 0.6476 - accuracy: 0.7793 - val_loss: 0.5576 - val_accuracy: 0.8116
Epoch 4/60
375/375 [==============================] - 6s 17ms/step - loss: 0.5926 - accuracy: 0.7969 - val_loss: 0.5255 - val_accuracy: 0.8216
Epoch 5/60
375/375 [==============================] - 6s 16ms/step - loss: 0.5588 - accuracy: 0.8096 - val_loss: 0.5021 - val_accuracy: 0.8297
Epoch 6/60
375/375 [==============================] - 6s 15ms/step - loss: 0.5347 - accuracy: 0.8160 - val_loss: 0.4841 - val_accuracy: 0.8358
Epoch 7/60
375/375 [==============================] - 5s 14ms/step - loss: 0.5134 - accuracy: 0.8234 - val_loss: 0.4705 - val_accuracy: 0.8392

# Calculate accuracy 

In [22]:
metrics = model.evaluate(x_test, y_test) #returns loss and accuracy
print(metrics[1])
print(f'Accuracy: {metrics[1]*100:.2f}%\n')

313/313 [==============================] - 1s 3ms/step - loss: 0.3466 - accuracy: 0.8761
0.8761000037193298
Accuracy: 87.61%



# Define Model (Drop rate : 0.5, No BN)

In [24]:
# for reproducibility
import random, os
os.environ['PYTHONHASHSEED']='0'
random.seed(123)
np.random.seed(123)
tf.random.set_seed(123)
sess = tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1,    inter_op_parallelism_threads=1,
                                                               allow_soft_placement=True, device_count = {'CPU': 1}))
from tensorflow.python.keras import backend as K
K.set_session(sess)



kernel_initializer='glorot_uniform'
activation_function = 'relu'

with tf.device('/cpu:0'):
    model = Sequential()
    model.add(Dense(512, activation='relu', input_shape=(784,)))
    model.add(Dropout(0.5))
    #model.add(BatchNormalization())
    model.add(Dense(512, activation='relu'))
    model.add(Dropout(0.5))
    #model.add(BatchNormalization())
    model.add(Dense(num_classes, activation='softmax'))

    model.summary()

    model.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 512)               401920    
                                                                 
 dropout_8 (Dropout)         (None, 512)               0         
                                                                 
 dense_13 (Dense)            (None, 512)               262656    
                                                                 
 dropout_9 (Dropout)         (None, 512)               0         
                                                                 
 dense_14 (Dense)            (None, 10)                5130      
                                                                 
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________


# Start Training

In [26]:
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 6s 14ms/step - loss: 1.4219 - accuracy: 0.5107 - val_loss: 0.8140 - val_accuracy: 0.7207
Epoch 2/60
375/375 [==============================] - 5s 13ms/step - loss: 0.9017 - accuracy: 0.6831 - val_loss: 0.6748 - val_accuracy: 0.7586
Epoch 3/60
375/375 [==============================] - 5s 13ms/step - loss: 0.7783 - accuracy: 0.7286 - val_loss: 0.6099 - val_accuracy: 0.7836
Epoch 4/60
375/375 [==============================] - 5s 12ms/step - loss: 0.7092 - accuracy: 0.7524 - val_loss: 0.5697 - val_accuracy: 0.7999
Epoch 5/60
375/375 [==============================] - 5s 13ms/step - loss: 0.6586 - accuracy: 0.7706 - val_loss: 0.5366 - val_accuracy: 0.8120
Epoch 6/60
375/375 [==============================] - 5s 13ms/step - loss: 0.6228 - accuracy: 0.7841 - val_loss: 0.5160 - val_accuracy: 0.8202
Epoch 7/60
375/375 [==============================] - 5s 13ms/step - loss: 0.5973 - accuracy: 0.7930 - val_loss: 0.4993 - val_accuracy: 0.8239

In [27]:
# Calculate accuracy 

In [28]:
metrics = model.evaluate(x_test, y_test) #returns loss and accuracy
print(metrics[1])
print(f'Accuracy: {metrics[1]*100:.2f}%\n')

313/313 [==============================] - 1s 3ms/step - loss: 0.3614 - accuracy: 0.8698
0.8697999715805054
Accuracy: 86.98%



# Define Model (Drop rate : 0.8, No BN)

In [34]:
# for reproducibility
import random, os
os.environ['PYTHONHASHSEED']='0'
random.seed(123)
np.random.seed(123)
tf.random.set_seed(123)
sess = tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1,    inter_op_parallelism_threads=1,
                                                               allow_soft_placement=True, device_count = {'CPU': 1}))
from tensorflow.python.keras import backend as K
K.set_session(sess)



kernel_initializer='glorot_uniform'
activation_function = 'relu'

with tf.device('/cpu:0'):
    model = Sequential()
    model.add(Dense(512, activation='relu', input_shape=(784,)))
    model.add(Dropout(0.8))
    #model.add(BatchNormalization())
    model.add(Dense(512, activation='relu'))
    model.add(Dropout(0.8))
    #model.add(BatchNormalization())
    model.add(Dense(num_classes, activation='softmax'))

    model.summary()

    model.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_18 (Dense)            (None, 512)               401920    
                                                                 
 dropout_12 (Dropout)        (None, 512)               0         
                                                                 
 dense_19 (Dense)            (None, 512)               262656    
                                                                 
 dropout_13 (Dropout)        (None, 512)               0         
                                                                 
 dense_20 (Dense)            (None, 10)                5130      
                                                                 
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________


# Start Training

In [36]:
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 6s 14ms/step - loss: 1.9859 - accuracy: 0.2964 - val_loss: 1.2486 - val_accuracy: 0.6500
Epoch 2/60
375/375 [==============================] - 5s 13ms/step - loss: 1.3898 - accuracy: 0.4927 - val_loss: 0.9015 - val_accuracy: 0.6891
Epoch 3/60
375/375 [==============================] - 5s 12ms/step - loss: 1.1610 - accuracy: 0.5699 - val_loss: 0.7837 - val_accuracy: 0.7177
Epoch 4/60
375/375 [==============================] - 7s 17ms/step - loss: 1.0343 - accuracy: 0.6176 - val_loss: 0.7243 - val_accuracy: 0.7370
Epoch 5/60
375/375 [==============================] - 5s 14ms/step - loss: 0.9531 - accuracy: 0.6453 - val_loss: 0.6813 - val_accuracy: 0.7552
Epoch 6/60
375/375 [==============================] - 4s 12ms/step - loss: 0.8955 - accuracy: 0.6671 - val_loss: 0.6515 - val_accuracy: 0.7626
Epoch 7/60
375/375 [==============================] - 5s 14ms/step - loss: 0.8533 - accuracy: 0.6866 - val_loss: 0.6255 - val_accuracy: 0.7742

# Calculate accuracy 

In [38]:
metrics = model.evaluate(x_test, y_test) #returns loss and accuracy
print(metrics[1])
print(f'Accuracy: {metrics[1]*100:.2f}%\n')

313/313 [==============================] - 1s 2ms/step - loss: 0.4210 - accuracy: 0.8480
0.8479999899864197
Accuracy: 84.80%



# Define Model (Drop rate : 0.2, with BN)

In [40]:
# for reproducibility
import random, os
os.environ['PYTHONHASHSEED']='0'
random.seed(123)
np.random.seed(123)
tf.random.set_seed(123)
sess = tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1,    inter_op_parallelism_threads=1,
                                                               allow_soft_placement=True, device_count = {'CPU': 1}))
from tensorflow.python.keras import backend as K
K.set_session(sess)



kernel_initializer='glorot_uniform'
activation_function = 'relu'

with tf.device('/cpu:0'):
    model = Sequential()
    model.add(Dense(512, activation='relu', input_shape=(784,)))
    model.add(Dropout(0.2))
    model.add(BatchNormalization())
    model.add(Dense(512, activation='relu'))
    model.add(Dropout(0.2))
    model.add(BatchNormalization())
    model.add(Dense(num_classes, activation='softmax'))

    model.summary()

    model.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_21 (Dense)            (None, 512)               401920    
                                                                 
 dropout_14 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization (BatchN  (None, 512)              2048      
 ormalization)                                                   
                                                                 
 dense_22 (Dense)            (None, 512)               262656    
                                                                 
 dropout_15 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_1 (Batc  (None, 512)              2048      
 hNormalization)                                      

# Start Training

In [42]:
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 6s 15ms/step - loss: 0.7245 - accuracy: 0.7479 - val_loss: 0.5080 - val_accuracy: 0.8260
Epoch 2/60
375/375 [==============================] - 5s 14ms/step - loss: 0.5077 - accuracy: 0.8201 - val_loss: 0.4144 - val_accuracy: 0.8537
Epoch 3/60
375/375 [==============================] - 5s 14ms/step - loss: 0.4550 - accuracy: 0.8367 - val_loss: 0.3863 - val_accuracy: 0.8639
Epoch 4/60
375/375 [==============================] - 5s 15ms/step - loss: 0.4227 - accuracy: 0.8490 - val_loss: 0.3698 - val_accuracy: 0.8681
Epoch 5/60
375/375 [==============================] - 5s 14ms/step - loss: 0.4017 - accuracy: 0.8554 - val_loss: 0.3604 - val_accuracy: 0.8702
Epoch 6/60
375/375 [==============================] - 5s 14ms/step - loss: 0.3839 - accuracy: 0.8629 - val_loss: 0.3506 - val_accuracy: 0.8731
Epoch 7/60
375/375 [==============================] - 7s 18ms/step - loss: 0.3695 - accuracy: 0.8669 - val_loss: 0.3431 - val_accuracy: 0.8769

# Calculate accuracy 

In [44]:
metrics = model.evaluate(x_test, y_test) #returns loss and accuracy
print(metrics[1])
print(f'Accuracy: {metrics[1]*100:.2f}%\n')

313/313 [==============================] - 1s 3ms/step - loss: 0.3453 - accuracy: 0.8855
0.8855000138282776
Accuracy: 88.55%



# Define Model (Drop rate : 0.5, with BN)

In [46]:
# for reproducibility
import random, os
os.environ['PYTHONHASHSEED']='0'
random.seed(123)
np.random.seed(123)
tf.random.set_seed(123)
sess = tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1,    inter_op_parallelism_threads=1,
                                                               allow_soft_placement=True, device_count = {'CPU': 1}))
from tensorflow.python.keras import backend as K
K.set_session(sess)



kernel_initializer='glorot_uniform'
activation_function = 'relu'

with tf.device('/cpu:0'):
    model = Sequential()
    model.add(Dense(512, activation='relu', input_shape=(784,)))
    model.add(Dropout(0.5))
    model.add(BatchNormalization())
    model.add(Dense(512, activation='relu'))
    model.add(Dropout(0.5))
    model.add(BatchNormalization())
    model.add(Dense(num_classes, activation='softmax'))

    model.summary()

    model.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_24 (Dense)            (None, 512)               401920    
                                                                 
 dropout_16 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_2 (Batc  (None, 512)              2048      
 hNormalization)                                                 
                                                                 
 dense_25 (Dense)            (None, 512)               262656    
                                                                 
 dropout_17 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_3 (Batc  (None, 512)              2048      
 hNormalization)                                      

# Start Training

In [48]:
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 6s 15ms/step - loss: 1.0639 - accuracy: 0.6358 - val_loss: 0.6040 - val_accuracy: 0.7910
Epoch 2/60
375/375 [==============================] - 5s 14ms/step - loss: 0.6905 - accuracy: 0.7562 - val_loss: 0.4847 - val_accuracy: 0.8267
Epoch 3/60
375/375 [==============================] - 5s 15ms/step - loss: 0.6121 - accuracy: 0.7832 - val_loss: 0.4523 - val_accuracy: 0.8390
Epoch 4/60
375/375 [==============================] - 5s 15ms/step - loss: 0.5653 - accuracy: 0.7994 - val_loss: 0.4338 - val_accuracy: 0.8446
Epoch 5/60
375/375 [==============================] - 5s 14ms/step - loss: 0.5367 - accuracy: 0.8099 - val_loss: 0.4155 - val_accuracy: 0.8517
Epoch 6/60
375/375 [==============================] - 5s 14ms/step - loss: 0.5125 - accuracy: 0.8190 - val_loss: 0.4058 - val_accuracy: 0.8540
Epoch 7/60
375/375 [==============================] - 5s 13ms/step - loss: 0.4924 - accuracy: 0.8261 - val_loss: 0.4006 - val_accuracy: 0.8563

# Calculate accuracy 

In [50]:
metrics = model.evaluate(x_test, y_test) #returns loss and accuracy
print(metrics[1])
print(f'Accuracy: {metrics[1]*100:.2f}%\n')

313/313 [==============================] - 1s 3ms/step - loss: 0.3334 - accuracy: 0.8816
0.881600022315979
Accuracy: 88.16%



# Define Model (Drop rate : 0.8, with BN)

In [52]:
# for reproducibility
import random, os
os.environ['PYTHONHASHSEED']='0'
random.seed(123)
np.random.seed(123)
tf.random.set_seed(123)
sess = tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1,    inter_op_parallelism_threads=1,
                                                               allow_soft_placement=True, device_count = {'CPU': 1}))
from tensorflow.python.keras import backend as K
K.set_session(sess)



kernel_initializer='glorot_uniform'
activation_function = 'relu'

with tf.device('/cpu:0'):
    model = Sequential()
    model.add(Dense(512, activation='relu', input_shape=(784,)))
    model.add(Dropout(0.8))
    model.add(BatchNormalization())
    model.add(Dense(512, activation='relu'))
    model.add(Dropout(0.8))
    model.add(BatchNormalization())
    model.add(Dense(num_classes, activation='softmax'))

    model.summary()

    model.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_27 (Dense)            (None, 512)               401920    
                                                                 
 dropout_18 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_4 (Batc  (None, 512)              2048      
 hNormalization)                                                 
                                                                 
 dense_28 (Dense)            (None, 512)               262656    
                                                                 
 dropout_19 (Dropout)        (None, 512)               0         
                                                                 
 batch_normalization_5 (Batc  (None, 512)              2048      
 hNormalization)                                      

# Start Training

In [54]:
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=0.2)

Epoch 1/60
375/375 [==============================] - 6s 14ms/step - loss: 2.0191 - accuracy: 0.3192 - val_loss: 1.0255 - val_accuracy: 0.6790
Epoch 2/60
375/375 [==============================] - 5s 14ms/step - loss: 1.3027 - accuracy: 0.5304 - val_loss: 0.7753 - val_accuracy: 0.7373
Epoch 3/60
375/375 [==============================] - 5s 15ms/step - loss: 1.0819 - accuracy: 0.6018 - val_loss: 0.7039 - val_accuracy: 0.7505
Epoch 4/60
375/375 [==============================] - 6s 15ms/step - loss: 0.9767 - accuracy: 0.6427 - val_loss: 0.6691 - val_accuracy: 0.7624
Epoch 5/60
375/375 [==============================] - 5s 14ms/step - loss: 0.8964 - accuracy: 0.6703 - val_loss: 0.6348 - val_accuracy: 0.7768
Epoch 6/60
375/375 [==============================] - 5s 14ms/step - loss: 0.8482 - accuracy: 0.6897 - val_loss: 0.6112 - val_accuracy: 0.7829
Epoch 7/60
375/375 [==============================] - 5s 13ms/step - loss: 0.8080 - accuracy: 0.7076 - val_loss: 0.5990 - val_accuracy: 0.7896

# Calculate accuracy 

In [56]:
metrics = model.evaluate(x_test, y_test) #returns loss and accuracy
print(metrics[1])
print(f'Accuracy: {metrics[1]*100:.2f}%\n')

313/313 [==============================] - 3s 8ms/step - loss: 0.4243 - accuracy: 0.8516
0.8515999913215637
Accuracy: 85.16%



# 분석
최종 결과는 다음과 같다.
(BN은 Batch Normalization을 나타냄)

- Drop rate = 0.2, no BN : 87.61%
- Drop rate = 0.5, no BN : 86.98%
- Drop rate = 0.8, no BN : 84.80%
- Drop rate = 0.2, with BN : 88.55%
- Drop rate = 0.5, with BN : 88.16%
- Drop rate = 0.8, with BN : 85.16%

Drop rate는 해당 확률로 뉴런을 제거를 하겠다는 것을 의미한다.
과적합 방지를 위해 이러한 기법을 사용한다.
전체적으로 결과를 봤을 때 BN을 적용한 결과에서나 적용하지 않은 결과에서나
Drop rate가 높아질수록 정확도가 떨어지는 것을 확인할 수 있다.
이는 너무 높은 확률로 drop을 하게 되면 과소 학습된 결과가 나올 수 있기 때문이다.
또한 같은 drop rate에 대해서는
BN(Batch Normalization)을 사용한 결과가 전체적으로 더 정확도가 높은 것을 확인할 수 있다.
이는 BN을 사용하게 되면
layer 각 층마다 활성화 값을 normalize하기 때문에
초기값 선택의 영향을 적게 받게 되기 때문이다.
또한 기울기 손실(Gradient Vanishing)문제도 해결이 되기 때문에
정확도가 더 높아진다.